### PR TITLE
refactor: return OpResult in DbSlice::AddOrFind instead of throwing std::bad_alloc

### PR DESCRIPTION
**Pull request by kostasrim**
resolve comments in #2379

* return OpResult in AddOrFind instead of throwing bad_alloc

### CODE REVIEW

File: src/server/db_slice.cc [hunk 1]
```diff
@@ -495,7 +495,8 @@ OpResult<pair<PrimeConstIterator, unsigned>> DbSlice::FindFirstReadOnly(const Co
   return OpStatus::KEY_NOTFOUND;
 }
 
-DbSlice::AddOrFindResult DbSlice::AddOrFind(const Context& cntx, string_view key) noexcept(false) {
+OpResult<DbSlice::AddOrFindResult> DbSlice::AddOrFind(const Context& cntx,
+                                                      string_view key) noexcept(false) {
```
Comment (chakaz) [inline src/server/db_slice.cc:499]: can we change this to `noexcept(true)` now? or just `noexcept`?
Or just remove this, as we never throw in other places anyway?

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -774,12 +776,18 @@ uint32_t DbSlice::GetMCFlag(DbIndex db_ind, const PrimeKey& key) const {
   return it->second;
 }
 
-DbSlice::ItAndUpdater DbSlice::AddNew(const Context& cntx, string_view key, PrimeValue obj,
-                                      uint64_t expire_at_ms) noexcept(false) {
-  auto res = AddOrUpdateInternal(cntx, key, std::move(obj), expire_at_ms, false);
+OpResult<DbSlice::ItAndUpdater> DbSlice::AddNew(const Context& cntx, string_view key,
+                                                PrimeValue obj,
+                                                uint64_t expire_at_ms) noexcept(false) {
```
Comment (chakaz) [inline src/server/db_slice.cc:781]: same question here

File: src/server/db_slice.cc [hunk 3]
```diff
@@ -832,14 +840,21 @@ OpResult<int64_t> DbSlice::UpdateExpire(const Context& cntx, PrimeIterator prime
...
 }
 
-DbSlice::AddOrFindResult DbSlice::AddOrUpdateInternal(const Context& cntx, std::string_view key,
-                                                      PrimeValue obj, uint64_t expire_at_ms,
-                                                      bool force_update) noexcept(false) {
+OpResult<DbSlice::AddOrFindResult> DbSlice::AddOrUpdateInternal(const Context& cntx,
+                                                                std::string_view key,
+                                                                PrimeValue obj,
+                                                                uint64_t expire_at_ms,
+                                                                bool force_update) noexcept(false) {
```
Comment (chakaz) [inline src/server/db_slice.cc:847]: ditto, and in other places :)

File: src/server/generic_family.cc [hunk 4]
```diff
@@ -1491,7 +1498,11 @@ OpStatus GenericFamily::OpMove(const OpArgs& op_args, string_view key, DbIndex t
   from_res.it->second.SetExpire(IsValid(from_res.exp_it));
 
   CHECK(db_slice.Del(op_args.db_cntx.db_index, from_res.it));
-  auto add_res = db_slice.AddNew(target_cntx, key, std::move(from_obj), exp_ts);
+  auto op_result = db_slice.AddNew(target_cntx, key, std::move(from_obj), exp_ts);
+  if (!op_result) {
+    return op_result.status();
+  }
+  auto add_res = std::move(*op_result);
```
Comment (chakaz) [inline src/server/generic_family.cc:1505]:
```suggestion
  auto* add_res = *op_result;
```

File: src/server/hll_family.cc [hunk 5]
```diff
@@ -70,33 +70,33 @@ OpResult<int> AddToHll(const OpArgs& op_args, string_view key, CmdArgList values
...
-      return OpStatus::WRONG_TYPE;
-    } else {
-      res.it->second.GetString(&hll);
-      ConvertToDenseIfNeeded(&hll);
-    }
+  auto op_res = db_slice.AddOrFind(op_args.db_cntx, key);
+  if (!op_res) {
+    return op_res.status();
+  }
+  auto res = std::move(*op_res);
```
Comment (chakaz) [inline src/server/hll_family.cc:77]:
```suggestion
  auto& res = *op_res;
```

File: src/server/hll_family.cc [hunk 6]
```diff
@@ -258,12 +258,11 @@ OpResult<int> PFMergeInternal(CmdArgList args, ConnectionContext* cntx) {
...
-    DbSlice::AddOrFindResult res;
-    try {
-      res = db_slice.AddOrFind(t->GetDbContext(), key);
-    } catch (const bad_alloc& e) {
-      return OpStatus::OUT_OF_MEMORY;
+    auto op_res = db_slice.AddOrFind(t->GetDbContext(), key);
+    if (!op_res) {
+      return op_res.status();
     }
+    auto res = std::move(*op_res);
```
Comment (chakaz) [inline src/server/hll_family.cc:265]:
```suggestion
    auto& res = *op_res;
```

File: src/server/hset_family.cc [hunk 7]
```diff
@@ -168,12 +168,11 @@ OpStatus IncrementValue(optional<string_view> prev_val, IncrByParam* param) {
...
-  DbSlice::AddOrFindResult add_res;
-  try {
-    add_res = db_slice.AddOrFind(op_args.db_cntx, key);
-  } catch (const bad_alloc& e) {
-    return OpStatus::OUT_OF_MEMORY;
+  auto op_res = db_slice.AddOrFind(op_args.db_cntx, key);
+  if (!op_res) {
+    return op_res.status();
   }
+  auto add_res = std::move(*op_res);
```
Comment (chakaz) [inline src/server/hset_family.cc:175]:
```suggestion
  auto& add_res = *op_res;
```

Comment (chakaz) [reply src/server/hset_family.cc:175]: I'm tired of writing these, also in other places in this PR :)

File: src/server/db_slice.cc (hunk 1 above)
Comment (kostasrim) [reply src/server/db_slice.cc:499]: just `noexcept` will suffice but if we omit it completely it's also ok I guess

File: src/server/db_slice.cc [hunk 8]
```diff
@@ -536,33 +536,36 @@ OpResult<pair<PrimeConstIterator, unsigned>> DbSlice::FindFirstReadOnly(const Co
...
+        .exp_it = res->exp_it,
+        .is_new = false,
+        .post_updater = AutoUpdater({.action = AutoUpdater::DestructorAction::kRun,
+                                     .db_slice = this,
+                                     .db_ind = cntx.db_index,
+                                     .it = res->it,
+                                     .key = key})};
+  }
+  auto status = res.status();
+  CHECK_EQ(status == OpStatus::KEY_NOTFOUND || status == OpStatus::OUT_OF_MEMORY, true);
```
Comment (chakaz) [inline src/server/db_slice.cc:568]:
```suggestion
  CHECK(status == OpStatus::KEY_NOTFOUND || status == OpStatus::OUT_OF_MEMORY) << status;
```

File: src/server/dragonfly_test.cc [hunk 9]
```diff
@@ -541,19 +541,19 @@ TEST_F(DflyEngineTest, Bug496) {
         db.RegisterOnChange([&cb_hits](DbIndex, const DbSlice::ChangeReq&) { cb_hits++; });
 
     {
-      auto res = db.AddOrFind({}, "key-1");
+      auto res = std::move(*db.AddOrFind({}, "key-1"));
```
Comment (chakaz) [inline src/server/dragonfly_test.cc:544]: Here's a proposal:
Maybe add `V&& operator*() &&;` to `OpResult`, and then we won't have to call `std::move()` on what seems like rvalue types? We do it quite a lot, so I think it's worth it. That's also how `std::optional` allows similar things.

File: src/server/bitops_family.cc [hunk 10]
```diff
@@ -323,21 +323,20 @@ std::optional<bool> ElementAccess::Exists(EngineShard* shard) {
...
-    added_ = add_res.is_new;
-    shard_ = shard;
-    post_updater_ = std::move(add_res.post_updater);
-    return OpStatus::OK;
-  } catch (const std::bad_alloc&) {
-    return OpStatus::OUT_OF_MEMORY;
+  auto op_res = shard->db_slice().AddOrFindAndFetch(context_, key_);
+  if (!op_res) {
+    return op_res.status();
+  }
```
Comment (chakaz) [inline src/server/bitops_family.cc:329]: We have a similar `if` in many places. For `std::status_code` we have `RETURN_ON_ERR()` (see `error.h`), so how about creating a `RETURN_ON_BAD_STATUS()` macro?

**Review by chakaz — APPROVED** [review]
awesome :)

**Pull request merged**